internal/p10k.zsh

```diff
@@ -4280,24 +4281,59 @@ function instant_prompt_chezmoi_shell() {
 # Virtualenv: current working virtualenv
 # More information on virtualenv (Python):
 # https://virtualenv.pypa.io/en/latest/
+_virtualenv_load_name_VIRTUAL_ENV() {
+  # VIRTUAL_ENV is the on-disk path, VIRTUAL_ENV_PROMPT is `(name_of_virtualenv) `.
+  # Use custom name if it was set (python -m venv -p "foo" .venv)
+  if [[ $VIRTUAL_ENV_PROMPT == '('?*') ' && $VIRTUAL_ENV_PROMPT != "($n) " ]]; then
+    _virtualenv_name=$VIRTUAL_ENV_PROMPT[2,-3]
+  # Use parent directory name if virtualenv name is generic (eg .venv)
+  elif [[ ${VIRTUAL_ENV:t} == $~_POWERLEVEL9K_VIRTUALENV_GENERIC_NAMES ]]; then
+    _virtualenv_name=${VIRTUAL_ENV:h:t}
+  # Otherwise use the name as-is
+  else
+    _virtualenv_name=${VIRTUAL_ENV:t}
+  fi
+}
+_virtualenv_load_name_poetry() {
+  local idx=$1
+  local dir=${_p9k__parent_dirs[idx]}
+  local pyproject="$dir/pyproject.toml"
+  if ! _p9k_cached_cmd 0 "$pyproject" poetry -C "$dir" version; then
+    # poetry execution failed. Display nothing.
+    _virtualenv_name=""
+  fi
+  # Return the first word only, eg the value of pyproject.toml's `poetry.name`
+  _virtualenv_name="${_p9k__ret%% *}"
+}
 prompt_virtualenv() {
   local msg=''
+  local version=''
   if (( _POWERLEVEL9K_VIRTUALENV_SHOW_PYTHON_VERSION )) && _p9k_python_version; then
-    msg="${_p9k__ret//\%/%%} "
+    version="${_p9k__ret//\%/%%}"
+    msg="$version "
   fi
-  local v=${VIRTUAL_ENV:t}
-  if [[ $VIRTUAL_ENV_PROMPT == '('?*') ' && $VIRTUAL_ENV_PROMPT != "($v) " ]]; then
-    v=$VIRTUAL_ENV_PROMPT[2,-3]
-  elif [[ $v == $~_POWERLEVEL9K_VIRTUALENV_GENERIC_NAMES ]]; then
-    v=${VIRTUAL_ENV:h:t}
+  _virtualenv_name=''
+  if [[ -n $VIRTUAL_ENV ]]; then
+    _virtualenv_load_name_VIRTUAL_ENV
+  else
+    local start end
+    _p9k_upglob pyproject.toml
+    local idx=$?
+    if (( idx > 0 )); then
+      _virtualenv_load_name_poetry $idx
+    fi
+  fi
```

Comment:
It would be nice if this code had the structure

```zsh
_virtualenv_name=''
if in_virtual_env; then
  load_name_from_env_vars
elif in_poetry_project; then
  load_poetry_project_name
fi
```

The current structure is functionally the same, but harder to reason about.

Here's an alternate implementation. I'm not sure if it's better or worse, what do you think?

```zsh
_virtualenv_load_name_from_env_vars || _virtualenv_load_name_from_poetry || _virtualenv_name=''

_virtualenv_load_name_from_env_vars() {
  [[ -z ${VIRTUAL_ENV} ]] && return 1
  _virtualenv_name=${VIRTUAL_ENV}
}
_virtualenv_load_name_from_poetry() {
  _p9k_upglob pyproject.toml || return 1
  _p9k_cached_cmd 0 "$pyproject" poetry -C "$dir" version
  _virtualenv_name="${_p9k__ret%% *}"
}
```

Or do you have a better idea?